native/server/src/sqlite_brongnal.rs

```diff
@@ -0,0 +1,350 @@
+            )?,
+            ()
+        );
+        assert_eq!(
+            storage.get_current_keys("alice")?,
+            (alice_verifying_key, alice_spk)
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn get_keys_not_found() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        assert_eq!(
+            storage.get_current_keys("alice").err().map(|e| e.code()),
+            Some(Code::NotFound)
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn pop_empty_one_time_keys() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        assert_eq!(storage.pop_one_time_key("bob")?, None);
+        Ok(())
+    }
+
+    #[test]
+    fn retrieve_one_time_key() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        let mut bob = MemoryClient::new();
+        let keys = bob.add_one_time_keys(1)?.pre_keys;
+        storage.add_user(
+            String::from("bob"),
+            (&bob.get_identity_key()?).into(),
+            bob.get_spk()?.into(),
+        )?;
+        storage.add_one_time_keys("bob", keys.clone())?;
+        assert_eq!(storage.pop_one_time_key("bob")?, Some(keys[0]));
+        assert_eq!(storage.pop_one_time_key("bob")?, None);
+        Ok(())
+    }
+
+    #[test]
+    fn updating_pre_key_not_found() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        assert_eq!(
+            storage
+                .update_pre_key("bob", proto::service::SignedPreKey::default())
+                .err()
+                .map(|e| e.code()),
+            Some(Code::NotFound)
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn update_pre_key_success() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        let mut bob = MemoryClient::new();
+        let bob_verifying_key = VerifyingKey::from(&bob.get_identity_key().unwrap());
+        let mut bob_spk: proto::service::SignedPreKey = bob.get_spk().unwrap().into();
+        storage.add_user(String::from("bob"), bob_verifying_key, bob_spk.clone())?;
+
+        bob_spk.pre_key = Some(bob.add_one_time_keys(1)?.pre_keys[0].to_bytes().to_vec());
+        storage.update_pre_key("bob", bob_spk.clone())?;
+
+        assert_eq!(
+            storage.get_current_keys("bob")?,
+            (bob_verifying_key, bob_spk)
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn add_message_unknown_user() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
+        assert_eq!(
+            storage
+                .add_message("bob", proto::service::Message::default())
+                .err()
+                .map(|e| e.code()),
+            Some(Code::NotFound)
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn add_get_message() -> Result<()> {
+        let storage = SqliteStorage::new(Connection::open_in_memory()?)?;
```

Comment:
[nit] What's the runtime on creating a new connection each time in memory? Would it be significantly faster reusing a connection over the test and just removing cleaning the tables in-between?

Reply:
```
cargo test -- -Z unstable-options --report-time
running 8 tests
test sqlite_brongnal::tests::get_keys_not_found ... ok <0.003s>
test sqlite_brongnal::tests::updating_pre_key_not_found ... ok <0.003s>
test sqlite_brongnal::tests::add_message_unknown_user ... ok <0.004s>
test sqlite_brongnal::tests::pop_empty_one_time_keys ... ok <0.004s>
test sqlite_brongnal::tests::add_user_get_keys ... ok <0.005s>
test sqlite_brongnal::tests::add_get_message ... ok <0.006s>
test sqlite_brongnal::tests::retrieve_one_time_key ... ok <0.006s>
test sqlite_brongnal::tests::update_pre_key_success ... ok <0.007s>
```
Shared state between tests hides bugs.